### ISSUE TITLE
ReactMount.nodeCache is repopulated with purged nodes that are focused at unmount

### ISSUE THREAD

**Issue by syranide** [facebook/react#2988]
If you unmount a currently focused node, `nodeCache` is repopulated after being purged and unless that specific ID is revisited later, it will remain there forever. ~~This is mostly likely due to the focus/selection restoration phase after reconciliation.~~

Repro: http://dev.cetrez.com/jsx/nodecache.html (nodeCache is output into the console)

**Comment by syranide** [comment]
I'll go ahead and put *good first bug* on this one, it may be a bit tricky to find the root cause but it should be quite easy to narrow down and the fix should be rather simple too I think. So perhaps not the easiest bug if you're entirely new.

**Comment by javpaw** [comment]
Hi syranide, I'd like to take a look at this but the link is not working, can you check it please?

**Comment by syranide** [comment]
@javpaw I can have it fixed in an hour or so, but really, just change `var nodeCache = {}` to `var nodeCache = window.nodeCache = {}` and run that code against your own React build and you'll see it.

**Comment by syranide** [comment]
Fixed.

**Comment by pletcher** [comment]
After a bunch of experimentation related to the linked PR, I was able to trace this issue to `EventListener`, and specifically to the fact that event the default event handlers (added as part of `ReactBrowserEventEmitter.listenTo`) aren't removed when unmounting.

It looks like when an `<input>` is mounted, it attaches listeners for `focus`, `blur`, etc. 

https://github.com/facebook/react/blob/master/src/renderers/dom/client/ReactBrowserEventEmitter.js#L271-L275

```javascript
ReactBrowserEventEmitter.ReactEventListener.trapCapturedEvent(
  topLevelTypes.topBlur,
  'blur',
  mountAt
);
```

Which leads us to https://github.com/facebook/react/blob/master/src/renderers/dom/client/ReactEventListener.js#L145-L155

```javascript
trapCapturedEvent: function(topLevelType, handlerBaseName, handle) {
  var element = handle;
  if (!element) {
    return null;
  }  
  return EventListener.capture(
    element,
    handlerBaseName,
    ReactEventListener.dispatchEvent.bind(null, topLevelType)
  );
},
```

and finally to https://github.com/facebook/react/blob/master/src/shared/vendor/stubs/EventListener.js#L74-L79

```javascript
target.addEventListener(eventType, callback, true);
return {
  remove: function() {
    target.removeEventListener(eventType, callback, true);
   }
};
```

So a `remove` function is being returned, it's just not getting used. I think if we exposed `remove` as callable (and called it before the component is removed from the DOM), we'd be in the clear.

I'm not sure where best to handle this, given that I'm fairly new to React's source and it's sort of massive. Are there any conventions in use or places that I should look?

**Comment by spicyj** [comment]
@pletcher You're right that those aren't removed, but they're added only once per document at the top level so that shouldn't really be a problem.

**Comment by pletcher** [comment]
Right right, but that's what's firing the `blur` event that causes the node to get recached when it's unmounted (or at least, that's what the stack points to when I climb through it) -- the browser's native handling of removing an `input` component is triggering this handler.

**Comment by spicyj** [comment]
Yes; the fix is probably to change React's event handling to not fire events on removed nodes. (If another root is still mounted, we can't remove the top-level handler anyway.) cc #3298, #3790

**Comment by jimfb** [comment]
@spicyj I think we intentionally fire events on removed nodes.  I'm pretty sure there was a unit test for that, ensuring that we fire an event on nodes even if they are removed as part of an event handler as the event bubbles up.

**Comment by jimfb** [comment]
https://github.com/facebook/react/blob/master/src/renderers/dom/client/__tests__/ReactEventListener-test.js#L93

Spec is `it should not get confused by disappearing elements`, which I suppose is sufficiently vague as to imply a non-fatal but still undefined behavior :/.

**Comment by spicyj** [comment]
Yeah, goes back to https://github.com/facebook/react/issues/1105.

**Comment by bramschulting** [comment]
We encountered this problem as well, but we've found a dirty work-around. If you wrap `React.unmountComponentAtNode(target)` in a `setTimeout`, this will work.

We use this in a portal mixin, which renders a React component in a different DOM element. This raised another problem when rendering a new component directly after unmounting the first one. In that case, the second component won't show up. We fixed this by storing a unique key in the component state and the DOM element, and only calling `unmountComponentAtNode` if these keys are the same.

```
setTimeout(()=> {
  if (this.state.portalKey === this._target._portalKey) {
    React.unmountComponentAtNode(this._target);
  }
});
```

This works for now, but we'd love a less dirty, more native solution
